src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,256 @@
+            {
+                return 0;
+            }
+            numberOfBytes = Math.Min(numberOfBytes, buffer.Length);
+            buffer.Slice(0, numberOfBytes).CopyTo(new Span<byte>(buff, numberOfBytes));
+            Unsafe.Write(bio.ptr, buffer.Slice(numberOfBytes));
+            bio.num = buffer.Length - numberOfBytes;
+            return numberOfBytes;
+        }
+
+        public static void SetReadBufferPointer(InteropBio.BioHandle bio, ref ReadableBuffer buffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].num = buffer.Length;
+            b[0].ptr = Unsafe.AsPointer(ref buffer);
+        }
+
+        public static void SetWriteBufferPointer(InteropBio.BioHandle bio, GCHandle handleToWriteBuffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].ptr = (void*)GCHandle.ToIntPtr(handleToWriteBuffer);
+        }
+
+        private static long ControlBio(ref bio_st bio, BioControl cmd, long num, void* ptr)
+        {
+            switch (cmd)
+            {
+                case BioControl.BIO_CTRL_FLUSH:
+                case BioControl.BIO_CTRL_POP:
+                case BioControl.BIO_CTRL_PUSH:
+                    return 1;
+            }
+            return 0;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct bio_method_st
+        {
+            public int type;
+            public void* name;
+            public Write bwriteDelegate; // int (*bwrite) (BIO*, const char*, int);
+            public Read breadDelegate; //int (*bread) (BIO*, char*, int);
+            public void* bputsDelegate; //int (*bputs) (BIO*, const char*);
+            public void* bgestsDelegate; //int (*bgets) (BIO*, char*, int);
+            public Control ctrlDelegate; //long (*ctrl) (BIO*, int, long, void*);
+            public Create create; //int (*create) (BIO*);
+            public Free destroy; //int (*destroy) (BIO*);
+            public void* callback_ctrl; //long (*callback_ctrl) (BIO*, int, bio_info_cb*);
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        struct bio_st
+        {
+            public void* method;
+            /* bio, mode, argp, argi, argl, ret */
+            public void* callbackDelegate; //long (*callback) (struct bio_st *, int, const char*, int, long, long);
+            public byte* callBackArgsDelegate; //char* cb_arg;               /* first argument for the callback */
+            public int init;
+            public int shutdown;
+            public int flags;                  /* extra storage */
+            public int retry_reason;
+            public int num;
+            public void* ptr;
+            public void* next_bio;// struct bio_st *next_bio;    /* used by filter BIOs */
+            public void* prev_bio;//struct bio_st *prev_bio;    /* used by filter BIOs */
+            public int references;
+            public ulong num_read;
+            public ulong num_write;
+        };
```

Comment:
As far as I am aware the flow always for these Bio objects is. You give it a template of the function pointers. Then you can only allocate via a Bio_New call then you can interact with that created object.